### ISSUE TITLE
Flag indentation error without invalidating rest of file

### ISSUE THREAD

**Issue by parched** [microsoft/pylance-release#3656]
If there's some unexpected indentation it shows errors in other parts of the file (and any files that depend on it).

![image](https://user-images.githubusercontent.com/5975405/202940310-701a91bd-9d00-4003-9bf3-356b04bcd4c1.png)

This can make tracking down the real error more difficult. It would be great if the parser flagged the unexpected indentation but carried on parsing like that line didn't exist (alternatively as if it was indented correctly). Which would mean an error was only shown on line 7 above.

**Comment by parched** [comment]
Correcting the indentation rather than ignoring the line, would probably be a better choice, at least for an example like this
![image](https://user-images.githubusercontent.com/5975405/202943276-6d31a8e8-b148-4b5f-9a7d-bd2a20cb8b5f.png)

**Comment by parched** [comment]
Or, if it's under-indented, pretend there's an extra `pass` line with the correct indentation, and, if it's over-indented, pretend the indentation is decreased. I'm not really sure if this would work for all case though, or even if it's possible to come up with such a rule.

**Comment by erictraut** [comment]
Yes, we can do a better job of parse recovery here. It's a little more complicated than ignoring indents, but it's still pretty straightforward. In Python, the tokenizer inserts both `indent` and `dedent` tokens into the token stream. In your example above, the token stream consists of:
```
...
Indent (4)
Identifier (x)
Operator (=)
Literal Number (1)
New Line
Indent (8) <=== location of parse error: unexpected indent
Operator (+)
Literal Number (2)
New Line
Dedent (4) <=== parse recovery
Keyword (return)
Identifier (x)
Dedent (0)
```

The initial error occurs when the parser detects an unexpected indent. I've added better recovery logic when the next dedent token is encountered. The logic now determines whether the dedent leaves the indent amount less than that of the body of the function. If not, it assumes that it should continue to parse statements as though they're still part of the function.

This will be addressed in a future release of pyright & pylance.

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2022.11.41, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20221141-30-November-2022-PreRelease)
